sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -1935,23 +1935,91 @@ pi_result cuda_piProgramGetInfo(pi_program program, pi_program_info param_name,
   return {};
 }
 
-pi_result cuda_piProgramLink( // TODO: change interface to return error code
-    pi_context context, pi_uint32 num_devices, const pi_device *device_list,
-    const char *options, pi_uint32 num_input_programs,
-    const pi_program *input_programs,
-    void (*pfn_notify)(pi_program program, void *user_data), void *user_data,
-    pi_program *ret_program) {
-  cl::sycl::detail::pi::die("cuda_piProgramLink not implemented");
-  return {};
+pi_result cuda_piProgramLink(pi_context context, pi_uint32 num_devices,
+                             const pi_device *device_list, const char *options,
+                             pi_uint32 num_input_programs,
+                             const pi_program *input_programs,
+                             void (*pfn_notify)(pi_program program,
+                                                void *user_data),
+                             void *user_data, pi_program *ret_program) {
+
+  assert(ret_program != nullptr);
+  assert(num_devices == 1 || num_devices == 0);
+  assert(device_list != nullptr || num_devices == 0);
+  assert(pfn_notify == nullptr);
+  assert(user_data == nullptr);
+  pi_result retError = PI_SUCCESS;
+
+  try {
+    ScopedContext active(context);
+
+    CUlinkState state;
+    std::unique_ptr<_pi_program> retProgram{new _pi_program{context}};
+
+    // TODO: Linker options
+    retError = PI_CHECK_ERROR(cuLinkCreate(0, nullptr, nullptr, &state));
+    try {
+      for (size_t i = 0; i < num_input_programs; ++i) {
+        pi_program program = input_programs[i];
+        std::vector<char> source(program->source_,
```

Comment:
Why do you need temporary source ? Why can you just pass program->source_ ptr to cuLinkAddData?

Reply:
`cuLinkAddData` takes the source data as a `void *` which means we'd be scrapping the `const` modifier of `source_`. To avoid any problems we instead make a copy of `source_`.

Reply:
The documentation for cuLinkAddData says "Ownership of data is retained by the caller.", so it should be safe to pass original pointer. Current implementation looks like performance problem.

Reply:
Probably my reading of the documentation is incorrect.

Reply:
I could also be wrong. I'm testing it now and if it works we can assume that your quote is enough to assume that it won't try to alter the source. I do wonder though as to why they didn't go with `const void *` like they do for `cuModuleLoad`.

Reply:
It seems to be working without and since you are right that performance wise the previous approach was lackluster, I'm inclined to say that this is probably the better option. Though I'm still a bit concerned regarding the use of `void *` rather than `const void *`. Let's keep this in mind for the future if it starts breaking.